packages/ts/react-crud/src/autogrid.tsx

```diff
@@ -190,15 +190,41 @@ function useColumns(
     return (
       <ColumnContext.Provider
         key={propertyInfo.name}
-        value={{ propertyInfo, setPropertyFilter, sortState, setSortState, customColumnOptions }}
+        value={{
+          propertyInfo,
+          setPropertyFilter,
+          sortState,
+          setSortState,
+          customColumnOptions,
+        }}
       >
         {column}
       </ColumnContext.Provider>
     );
   });
 
   if (options.customColumns) {
-    columns = [...columns, ...options.customColumns];
+    if (options.visibleColumns) {
+      const columnMap = new Map<string, JSX.Element>();
+      columns.forEach((column) => {
+        const { key } = column;
+        if (key) {
+          columnMap.set(key, column);
+        }
+      });
+      options.customColumns.forEach((customColumn) => {
+        const { key } = customColumn;
+        if (key) {
+          columnMap.set(key, customColumn);
+        }
+      });
+      columns = Array(columns.length + options.customColumns.length).fill(null);
+      effectiveColumns.forEach((key, index) => {
+        columns[index] = columnMap.get(key)!;
+      });
```

Comment:
The reason behind choosing the Map was for not scanning either of those two arrays (columns and customColumns) for each effective column. Although, probably the JS Map structure is not be similar to a HashMap in java (and I think it scans and underlying array for the `get` operation), it might not be that big of an issue to choose either of the above suggested codes. I went for the Map version just with a hope that JS Map might be a better choice here semantically.